test/test_transformers.py

```diff
@@ -2143,6 +2143,72 @@ def test_scaled_dot_product_fused_attention_with_inf(self, device):
             actual = torch.nn.functional.scaled_dot_product_attention(q, k, v, attn_mask=mask)
         self.assertEqual(math_ref, actual)
 
+    @unittest.skipIf(not PLATFORM_SUPPORTS_FUSED_ATTENTION, "Fused SDPA was not built for this system")
+    @parametrize("backend", [SDPBackend.EFFICIENT_ATTENTION, SDPBackend.FLASH_ATTENTION])
+    @parametrize("seq_len", [32, 64, 128])
+    @parametrize("head_dim", [16, 32])
+    @parametrize("dtype", [torch.float32, torch.float16])
+    def test_fully_masked_out_rows(self, backend, device, seq_len, head_dim, dtype):
+        def attention_inputs(seq_len, head_dim, device, dtype, mask_probability=0.25):
+            query = torch.rand(1, 1, seq_len, head_dim, requires_grad=True, device=device, dtype=dtype)
+            key = torch.rand(1, 1, seq_len, head_dim, requires_grad=True, device=device, dtype=dtype)
+            value = torch.rand(1, 1, seq_len, head_dim, requires_grad=True, device=device, dtype=dtype)
+
+            # Create a mask with random row masking
+            mask = torch.ones(1, 1, seq_len, seq_len, dtype=torch.bool, device=device)
+
+            # Randomly select rows to mask
+            random_mask = torch.rand(seq_len, device=device) < mask_probability
+            mask[0, 0, random_mask, :] = False
+
+            # Randomly mask out some elements on along rows with 20% probz
+            mask = mask & (torch.rand(1, 1, seq_len, seq_len, device=device) < 0.2)
+
+            return query, key, value, mask
+
+
+        def compute_output_and_grads(query, key, value, mask, backend):
+            with sdpa_kernel(backend):
+                masked_out = scaled_dot_product_attention(query, key, value, attn_mask=mask)
+                loss = masked_out.sum()
+            grads = torch.autograd.grad(loss, [query, key, value])
+            return masked_out, grads
+
+        if backend == SDPBackend.FLASH_ATTENTION and "cuda" in str(device):
+            unittest.skip("FlashAttention does not support masks on cuda")
+            return
+        if backend == SDPBackend.EFFICIENT_ATTENTION and "cpu" in str(device):
+            unittest.skip("EfficientAttention does not support masks on cpu")
+            return
+        query, key, value, mask = attention_inputs(seq_len, head_dim, device, dtype)
+
+        # Compute results for the tested backend
+        backend_out, backend_grads = compute_output_and_grads(query, key, value, mask, backend)
+
+        # Compute results for the Math backend
+        math_out, math_grads = compute_output_and_grads(query, key, value, mask, SDPBackend.MATH)
```

Comment:
out of curiosity, since we're comparing against a math reference below, does this depend on the recent changes to use float32 precision for float16 math?

Reply:
The tolerances are the only thing that would rely on the updates to fp32 precision. This is more of a check that the updates to the fused kernels indeed reflects the new `_safe_softmax`

Reply:
sweet. depending on how the NaN -> zero conversion story ends up, I'll propose a test with input NaNs (not from fully masked-out rows) comparing fused and math kernels to ensure consistency in how those are handled